kubernetes-client/src/main/java/io/fabric8/kubernetes/client/BaseClient.java

```diff
@@ -102,38 +104,68 @@ public String getNamespace() {
     return namespace;
   }
 
-  public void setAdaptableOverride(Predicate<Class<?>> adaptableOverride) {
-    this.adaptableOverride = adaptableOverride;
+  public void setMatchingGroupPredicate(Predicate<String> unsupportedApiGroups) {
+    this.matchingGroupPredicate = unsupportedApiGroups;
+  }
+
+  public Predicate<String> getMatchingGroupPredicate() {
+    return matchingGroupPredicate;
   }
 
   @Override
-  public <C> Boolean isAdaptable(Class<C> type) {
-    if (type.isAssignableFrom(this.getClass())) {
-      return true;
+  public boolean hasApiGroup(String apiGroup, boolean exact) {
+    if (matchingGroupPredicate != null) {
+      return matchingGroupPredicate.test(apiGroup);
     }
-    ExtensionAdapter<C> adapter = Adapters.get(type);
-    if (adapter != null) {
-      if (adaptableOverride != null) {
-        return adaptableOverride.test(type);
-      }
-      return adapter.isAdaptable(this);
-    } else {
+    if (exact) {
+      return getApiGroup(apiGroup) != null;
+    }
+    APIGroupList apiGroups = getApiGroups();
+    if (apiGroups == null) {
       return false;
     }
+    return apiGroups
+        .getGroups()
+        .stream()
+        .anyMatch(g -> g.getName().endsWith(apiGroup));
   }
 
   @Override
-  public <C> C adapt(Class<C> type) {
+  public <C extends Client> Boolean isAdaptable(Class<C> type) {
+    // if type is an instanceof SupportTestingClient, then it's a proper
+    // test, otherwise it could be legacy support on an extension client
+    C toTest = adapt(type);
+    if (toTest instanceof SupportTestingClient) {
+      return ((SupportTestingClient) toTest).isSupported();
+    }
+    return true;
+  }
```

Comment:
I'm confused. My assumption was that isAdatpable in v6.0 would return true as long as a specific client implementation for that client class exists:

For example:
- `false` if openshift-client-api dependency is present, openshift-client isn't and method is invoked with client.isAdaptable(OpenShiftClient.class)

However, here I'm unsure if we are also checking isSupported once the client is adapted.

Reply:
adapt will no longer perform any checks - previously at least for the OpenShift it did, which caused problems in mock scenarios.

isAdaptable I'm proposing to deprecate, but leave functionally the same - for select extensions it will call isSupported, which will do a check of the apigroups (which used to be done against the rootapis).

5.x style:
```
if (client.isAdaptable(OpenShiftClient.class)) {
  client.adapt(OpenShiftClient.class);
  ...
}
```
That will still function the same with these changes.

6.x style:
```
OpenShiftClient osclient = client.adapt(OpenShiftClient.class);
if (osclient.isSupported()) {
  
}
```
Beyond that you are supposed to use:
```
client.supports(SomeResource.class); 
```
To determine down to a type level what is supported.

Reply:
> To determine down to a type level what is supported.

I should point out that we are effectively caching the results of those calls when the type is supported, because it backs into the same logic as client.resources(SomeResource.class) - and we never timeout those entries.

Reply:
> I should point out that we are effectively caching the results of those calls when the type is supported, because it backs into the same logic as client.resources(SomeResource.class) - and we never timeout those entries.

Yes, I saw that.

My question was more regarding a _newer_ behavior for the `isAdaptable` method. So in the case I'm pointing out, what will be the result (I adapt to a non-existing client implementation ---> `client.adapt(OpenShiftClient.class`)

Reply:
If you have the api dependency, but not the client dependency for the OpenShiftClient, both a call to adapt and isAdaptable will now throw an illegalargumentexception.  That is different than before in that isAdaptable would have returned false.  

Beyond this, with more work beyond #3966 it's possible that we can merge the openshift api and client back together as it could be written to just depend upon the client.